provider/dagprovider.go

```
@@ -0,0 +1,58 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/ipfs/boxo/fetcher"
+	fetcherhelpers "github.com/ipfs/boxo/fetcher/helpers"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-cidutil"
+	cidlink "github.com/ipld/go-ipld-prime/linking/cid"
+)
+
+// NewDAGProvider returns a provider that traverses a DAG from a root CID
+func NewDAGProvider(root cid.Cid, fetchConfig fetcher.Factory) KeyChanFunc {
+	return func(ctx context.Context) (<-chan cid.Cid, error) {
+		if root.String() == "" {
+			return nil, fmt.Errorf("root CID cannot be empty")
+		}
+		if fetchConfig == nil {
+			return nil, fmt.Errorf("fetcher config cannot be nil")
+		}
```

You can remove. Let it panic

---

Done